sled-agent/src/sim/instance.rs

```diff
@@ -47,33 +76,109 @@ impl Simulatable for SimInstance {
 
     fn request_transition(
         &mut self,
-        target: &InstanceRuntimeStateRequested,
+        target: &InstanceStateRequested,
     ) -> Result<Option<InstanceAction>, Error> {
-        self.state.request_transition(target)
-    }
-
-    fn execute_desired_transition(&mut self) -> Option<InstanceAction> {
-        if matches!(self.state.current().run_state, InstanceState::Rebooting) {
-            self.state.observe_transition(&PropolisInstanceState::Starting)
-        } else if let Some(desired) = self.state.desired() {
-            // These operations would typically be triggered via responses from
-            // Propolis, but for a simulated sled agent, this does not exist.
-            //
-            // Instead, we make transitions to new states based entirely on the
-            // value of "desired".
-            let observed = match desired.run_state {
-                InstanceStateRequested::Running => {
-                    PropolisInstanceState::Running
+        match target {
+            InstanceStateRequested::Running => match self.terminal_state() {
+                ApiInstanceState::Creating => {
+                    // The non-simulated sled agent explicitly and synchronously
+                    // publishes the "Starting" state when cold-booting a new VM
+                    // (so that the VM appears to be starting while its Propolis
+                    // process is being launched).
+                    self.state.transition(ApiInstanceState::Starting);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Running);
                 }
-                InstanceStateRequested::Stopped => {
-                    PropolisInstanceState::Stopped
+                ApiInstanceState::Starting
+                | ApiInstanceState::Running
+                | ApiInstanceState::Rebooting
+                | ApiInstanceState::Migrating => {}
+                ApiInstanceState::Stopping
+                | ApiInstanceState::Stopped
+                | ApiInstanceState::Failed
+                | ApiInstanceState::Destroyed => {
```

Comment:
Should we allow Failed / Destroyed instances to be "restarted" this way? Doesn't this imply behavior that could never happen in the non-simulated case?